pkg/s3wrapper/upload_iso.go

```diff
@@ -235,80 +194,247 @@ func (u *ISOUploader) getISOHeaderInfo(log logrus.FieldLogger, baseObjectName st
-	sourceStartBytes int64, sourceEndBytes int64) (*s3.CompletedPart, error) {
-	completedPartCopy, err := u.s3client.UploadPartCopy(&s3.UploadPartCopyInput{
-		Bucket:          aws.String(u.bucket),
-		Key:             aws.String(destObjectKey),
-		CopySource:      aws.String(fmt.Sprintf("/%s/%s", u.bucket, sourceObjectKey)),
-		CopySourceRange: aws.String(fmt.Sprintf("bytes=%d-%d", sourceStartBytes, sourceEndBytes)),
-		PartNumber:      aws.Int64(partNum),
-		UploadId:        uploadID,
+type multiUpload struct {
+	ctx             context.Context
+	log             logrus.FieldLogger
+	uploader        *ISOUploader
+	isoInfo         *isoInfo
+	origContents    *[]byte
+	wg              sync.WaitGroup
+	mutex           sync.Mutex
+	err             error
+	uploadID        string
+	sourceObjectKey string
+	destObjectKey   string
+	parts           completedParts
+}
+
+type chunk struct {
+	partNum          int64
+	sourceStartBytes int64
+	sourceEndBytes   int64
+}
+
+// completedParts is a wrapper to make parts sortable by their part number,
+// since S3 required this list to be sent in sorted order.
+type completedParts []*s3.CompletedPart
+
+func (a completedParts) Len() int           { return len(a) }
+func (a completedParts) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a completedParts) Less(i, j int) bool { return *a[i].PartNumber < *a[j].PartNumber }
+
+func (m *multiUpload) Upload(ignitionConfig string) error {
+	m.log.Debugf("Creating multi-part upload of ISO %s", m.destObjectKey)
+	multiOut, err := m.uploader.s3client.CreateMultipartUploadWithContext(
+		m.ctx, &s3.CreateMultipartUploadInput{Bucket: aws.String(m.uploader.bucket), Key: aws.String(m.destObjectKey)})
+	if err != nil {
+		err = errors.Wrapf(err, "Failed to start upload for %s", m.destObjectKey)
+		m.log.Error(err)
+		return err
+	}
+	m.uploadID = *multiOut.UploadId
+
+	m.log.Debugf("Starting goroutines copying chunks for ISO %s", m.destObjectKey)
+	maxChunks := int(m.isoInfo.baseObjectSize/copyPartChunkSizeBytes) + 1
+	ch := make(chan chunk, maxChunks)
+	for i := 0; i < maxChunks; i++ {
+		m.wg.Add(1)
+		go m.copyChunk(ch)
+	}
+
+	m.log.Debugf("Providing work for goroutines copying chunks for ISO %s", m.destObjectKey)
+	embeddedAreaPartNum := m.generateWorkForRange(ch, int64(1), 0, m.isoInfo.areaOffsetBytes-1)
+	m.generateWorkForRange(ch, embeddedAreaPartNum+1, m.isoInfo.areaOffsetBytes+minimumPartSizeBytes, m.isoInfo.baseObjectSize-1)
+	close(ch)
+
+	m.log.Debugf("Uploading embedded area (compressed ignition config) for ISO %s", m.destObjectKey)
+	err = m.uploadIgnition(m.log, embeddedAreaPartNum, ignitionConfig)
+	if err != nil {
+		return err
+	}
+	m.log.Debugf("Completed upload of embedded area for ISO %s, waiting for async copies", m.destObjectKey)
+
+	// We now finished sending the chunks to copy as well as uploading the embedded area, wait for copies to finish
+	m.wg.Wait()
+	m.complete()
```

Comment:
because it will not always be called Multipart will never be closed if there is an error in `uploadIgnition` for example.